# Importing Libraries

In [1]:
#Import libraries
import pandas as pd
import numpy as np
import os

In [2]:
path = r'C:\Users\praha\OneDrive\Documents\Python\02-2023 Instachart Basket Analysis'

In [3]:
path

'C:\\Users\\praha\\OneDrive\\Documents\\Python\\02-2023 Instachart Basket Analysis'

In [4]:
df_prods = pd.read_csv(os.path.join(path, '02. Data', 'Original Data', 'products.csv'), index_col = False)

In [5]:
df_ords = pd.read_csv(os.path.join(path, '02. Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

In [6]:
df_ords = df_ords.drop(columns = ['Unnamed: 0', 'eval_set'])

In [7]:
df_ords.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [8]:
#Direction 2: Interpreting the df.describe()
df_ords.describe()

order_id       user_id  order_number  orders_day_of_week  \
count  3.421083e+06  3.421083e+06  3.421083e+06        3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01        2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01        2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00        0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00        1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01        3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01        5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02        6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

All columns have the same number of records. The column 'order_number' has a mean of 17, minimum of 1, and maximum of 100. Since the median (2nd quartile = 11) is less than mean value (17), we can determine that the distribution of values for this column is right-skewed. By looking at the minimum and maximum values, we can determine that the column 'orders_day_of_week' has 7 values (7 days/week), the column 'order_hour_of_day' has 24 values (24 hours/day), the colum 'days_since_prior_order' has the minimum value of 0 and the maximum value of 30. Since column 'days_since_prior_order' has median (2nd quartile = 7) is less than mean (11), we can determine that the distribution of values for this column is right-skewed.       

In [9]:
#Direction 3 & 4: check for mixed-type data in df_ords
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

There is no mix-type data in orders data set

In [10]:
#Direction 5: Check for missing values in orders data set 
df_ords.isnull().sum()

order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

In [11]:
df_nan = df_ords[df_ords['days_since_prior_order'].isnull()== True]

In [12]:
df_nan

order_id  user_id  order_number  orders_day_of_week  \
0         2539329        1             1                   2   
11        2168274        2             1                   2   
26        1374495        3             1                   1   
39        3343014        4             1                   6   
45        2717275        5             1                   3   
...           ...      ...           ...                 ...   
3420930    969311   206205             1                   4   
3420934   3189322   206206             1                   3   
3421002   2166133   206207             1                   6   
3421019   2227043   206208             1                   1   
3421069   3154581   206209             1                   3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 6 columns]

In [13]:
df_ords.shape

(3421083, 6)

There are 206209 missing values of 3421083 records in column 'day_since_prior_order

In [14]:
#Direction 6: address the missing values
#median = 2nd Quartile = 7
df_ords['days_since_prior_order'].fillna(7, inplace = True)

In [15]:
df_ords

order_id  user_id  order_number  orders_day_of_week  \
0         2539329        1             1                   2   
1         2398795        1             2                   3   
2          473747        1             3                   3   
3         2254736        1             4                   4   
4          431534        1             5                   4   
...           ...      ...           ...                 ...   
3421078   2266710   206209            10                   5   
3421079   1854736   206209            11                   4   
3421080    626363   206209            12                   1   
3421081   2977660   206209            13                   1   
3421082    272231   206209            14                   6   

         order_hour_of_day  days_since_prior_order  
0                        8                     7.0  
1                        7                    15.0  
2                       12                    21.0  
3                        7                    29.0  
4                       15                    28.0  
...                    ...                     ...  
3421078                 18                    29.0  
3421079                 10                    30.0  
3421080                 12                    18.0  
3421081                 12                     7.0  
3421082                 14                    30.0  

[3421083 rows x 6 columns]

In [16]:
#check the missing values after addressing
df_ords.isnull().sum()

order_id                  0
user_id                   0
order_number              0
orders_day_of_week        0
order_hour_of_day         0
days_since_prior_order    0
dtype: int64

I chose to adrress missing value with median because the distribution of values for this column is skewed to the right, so that we can not use the mean to impute the missing values because the mean is influenced by extreme values (such as: 30).

In [17]:
df_dups = df_ords[df_ords.duplicated()]

In [18]:
#Direction 7 & 8: Check & address the duplicate of orders data set 
df_dups

Empty DataFrame
Columns: [order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

There is no duplicate in this data set

In [19]:
#Direction 9: Export data
df_ords.to_csv(os.path.join(path, '02. Data', 'Prepared Data', 'orders_checked.csv'))